### ISSUE TITLE
ENH: Series.explode_range

### ISSUE THREAD

**Issue by JacekPliszka** [pandas-dev/pandas#39630]
#### Is your feature request related to a problem?

Yes. Often I have 2 columns: start and stop and what I want is instead single column with all values from start to stop. currently I can do it with explode but I need mostly int64 and for this is it is a bit slow.

#### Describe the solution you'd like

Series.explode_range method

```
def explode_range(
    self, stop, left_shift=0, right_shift=0, ignore_index: bool = False
) -> "Series":
    """
    Creates series from exploded ranges from this series to the ones given in stop

    Currently works only in int64
    left_shift and right_shift parameters are used to delay or hasten start - stop

    .. versionadded:: 1.2.x

    Parameters
    ----------
    stop: Series providing stop point
    left_shift: shift for start point - 0 includes it, 1 not
    right_shift: shift for end point - 0 excludes it, 1 includes
    ignore_index : bool, default False
        If True, the resulting index will be labeled 0, 1, …, n - 1.

    Returns
    -------
    Series
        Exploded ranges to rows; index will be duplicated for these rows.
```

#### API breaking implications

This is new feature but for int64 only for now

#### Describe alternatives you've considered

Using explode

#### Additional context

I will provide a PR with sample code

**Comment by MarcoGorelli** [comment]
Thanks @JacekPliszka - closing as not generic enough for pandas (see https://github.com/pandas-dev/pandas/pull/39631#issuecomment-774775504)

**Comment by JacekPliszka** [comment]
I believe it is very generic. The problem is that reviewers claim there is better way to do it but  I cannot find it anywhere...

The answers I've seen on StackOverflow in my opinion are not nearly as efficient as what I've proposed.

**Comment by jreback** [comment]
then show a concrete use case and benchmarks

**Comment by JacekPliszka** [comment]
```
events = pd.concat(10**5 *[events1]) 
%timeit events.apply(lambda x: range(x.start, x.stop), axis=1).explode()
4.54 s ± 345 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
%timeit explode_range(events.start, events.stop)
6.57 ms ± 361 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
I work with datasets up to 1000x bigger where difference is even more pronounced...

**Comment by jorisvandenbossche** [comment]
@JacekPliszka I think you can do something close (in performance) as the cython code using some numpy functionality. 

Using the small ``events1`` dataframe from https://github.com/pandas-dev/pandas/pull/39631#issuecomment-774773023:

```
In [40]: start = np.asarray(events1.start)

In [41]: stop = np.asarray(events1.stop)

In [42]: counts = stop - start

In [43]: counts_shifted = np.insert(counts, 0, 0)[:-1]

In [44]: counts
Out[44]: array([5, 4, 1])

In [45]: increments = np.repeat(np.ones_like(start), counts).cumsum() - np.repeat(counts_shifted.cumsum(), counts) - 1

In [47]: increments
Out[47]: array([0, 1, 2, 3, 4, 0, 1, 2, 3, 0])

In [49]: np.repeat(start, counts) + increments
Out[49]: array([10, 11, 12, 13, 14, 27, 28, 29, 30, 30])
```

This gives the same result (for this example at least) as the output shown there.

Testing on the larger example:

```
In [50]: events1 = pd.concat(10**5 *[events1])

In [51]: start = np.asarray(events1.start)

In [52]: stop = np.asarray(events1.stop)

In [53]: %%timeit
    ...: counts = stop - start
    ...: counts_shifted = np.insert(counts, 0, 0)[:-1]
    ...: increments = np.repeat(np.ones_like(start), counts).cumsum() - np.repeat(counts_shifted.cumsum(), counts) - 1
    ...: np.repeat(start, counts) + increments

18 ms ± 908 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Can you verify if this gives the desired result?

**Comment by jorisvandenbossche** [comment]
Some general notes:

- I think I agree that the proposed `explode_range` seems a bit too specific for inclusion in pandas itself *at this point* (we already have a huge API surface on the DataFrame/Series objects, so we need to put the bar high for new methods). At least not as a method (a utility function is something else IMO).
- Now, if we at some point gain support for a list type, *then* this might be useful as a creation method (to create a List column from  ranges). But as long as we don't have that (and thus can't store such data efficiently in a column without converting to python lists), I don't know if it makes sense to include such a method (especially if a numpy based method works for it)
- I follow @JacekPliszka in that this can be a quite typical use case, so I think adding an example like that to the docs would be welcome.

**Comment by JacekPliszka** [comment]
Yes, the result is as expected. I got 20ms on my PC but this is very good result.

The use case is a bit wider if you remember that period data is actually int64 - so it is like 'list of time periods on which something happened'

What do I need to do to have it as utility method?  Where shall I put it?
